test/unit/test_literal_lexer.ml

```diff
@@ -0,0 +1,90 @@
+open Base
+open Ocamlformat_lib
+
+let single_quote = '\''
+
+let double_quote = '"'
+
+let backslash = '\\'
+
+let space = ' '
+
+let tab = '\t'
+
+let newline = '\n'
+
+let tests_string =
+  let parse_string mode s =
+    let lexbuf = Lexing.from_string s in
+    match Literal_lexer.string mode lexbuf with
+    | r -> Some r
+    | exception _ -> None
+  in
+  let test_opt name s mode ~expected =
+    ( "string: " ^ name
+    , `Quick
+    , fun () ->
+        let got = parse_string mode s in
+        Alcotest.check Alcotest.(option string) Caml.__LOC__ expected got )
+  in
+  let test_one name s mode ~expected =
+    test_opt name s mode ~expected:(Some expected)
+  in
+  let test name s ~expected_preserve ~expected_normalize =
+    [ test_one (name ^ " (preserve)") s `Preserve ~expected:expected_preserve
+    ; test_one (name ^ " (normalize)") s `Normalize
+        ~expected:expected_normalize ]
+  in
+  List.concat
+    [ [test_opt "string: not a string" {|hello|} `Preserve ~expected:None]
+    ; test "simple" {|"hello"|} ~expected_preserve:"hello"
+        ~expected_normalize:"hello"
+    ; test "numeric escapes" {|"\123 \xff \o234"|}
+        ~expected_preserve:{|\123 \xff \o234|}
+        ~expected_normalize:{|\123 \xff \o234|}
+    ; test "raw tab"
+        (String.of_char_list [double_quote; tab; double_quote])
+        ~expected_preserve:(String.of_char_list [tab])
+        ~expected_normalize:(String.of_char_list [backslash; 't'])
+    ; test "backslash space"
+        (String.of_char_list [double_quote; backslash; space; double_quote])
+        ~expected_preserve:(String.of_char_list [backslash; space])
+        ~expected_normalize:(String.of_char_list [space])
+    ; test "backslash n"
+        (String.of_char_list [double_quote; backslash; 'n'; double_quote])
+        ~expected_preserve:(String.of_char_list [backslash; 'n'])
+        ~expected_normalize:(String.of_char_list [newline]) ]
+
+let tests_char =
+  let parse s =
+    let lexbuf = Lexing.from_string s in
+    match Literal_lexer.char lexbuf with r -> Some r | exception _ -> None
```

Comment:
Maybe we could replace the lexer exception by a result type? (not necessarily in this PR)

Reply:
it's defined in a mll file so I don't think we can change the API.

Reply:
Yes we can, we did it in mdx: https://github.com/realworldocaml/mdx/blob/master/lib/lexer_mdx.mll

Reply:
Ah nice, I'll try that!

Reply:
Done. I moved the lexbuf creation inside that function as well since it's always built from a string. how do that look?

Reply:
It looks better, thanks!

Reply:
thanks for the tip!